### DISCUSSION TITLE
How to use with Solid.js?

### DISCUSSION THREAD

**Discussion by thislooksfun** [oven-sh/bun#414] [Q&A] [answered]
Looking through the code and commits I see at least partial support for Solid.js, but I am unable to get `bun dev` to work with a starter Solid project.

Steps taken:

```sh
npx degit solidjs/templates/js my-app
cd my-app
rm pnpm-lock.yaml
bun install
bun dev
```

`bun dev` says it starts fine and is listening on `localhost:3000`, but when I open it in my browser it just shows a blank screen, and the console has a single error: `SyntaxError: Importing binding name 'template' is not found.`. I also tried calling `bun dev --jsx=solid` (and `bun dev --jsx solid`) and got the same results.

If the answer is that Solid just isn't supported yet, that's fine, but I got the impression from a [few of the commits](https://github.com/Jarred-Sumner/bun/search?q=solid&type=commits) that there was at least initial work towards Solid support, I just can't figure out how to make it work.

**Comment by tonyrewin** [comment]
Have you succeed unbounding bun from react in your template? 
`error: Could not resolve: "react/jsx-dev-runtime". Maybe you need to "bun install"?`
No, please, **bun** unbound yourself from react

```js
// Inserted by a transpiler (don't import it manually)
import {jsx as _jsx} from 'react/jsx-runtime';
```

**Comment by N0tExisting** [comment]
This should mostly depend on supporting `vite` + **`babel`**, as solid uses its own jsx-transform.

Here it seems like it may be resolving to the server build (only for SSR) instead of the dev/browser build as `template` not being found is a common symptom of that.  

> Note that solid also provides support for a regular `h`-style jsx-transform, however this is not recommended as it incurres a **heavy performance hit**.

**Reply by tonyrewin** [reply]
> may be resolving to the server build

this is how solidjs working, compiling components before runtime, no ssr [was](https://github.com/tonyrewin/solid-bun) used

**Reply by N0tExisting** [reply]
> this is how solidjs working, compiling components before runtime, no ssr [was](https://github.com/tonyrewin/solid-bun) used

# ?

It seems to resolve to the `node`(SSR) export condition instead of `dev`/`browser`

**Comment by kajetansw** [comment]
I also tried different command combinations and interestingly, when I did `bun dev --jsx-import-source=solid-js`, I got a different error 😄
```
Uncaught SyntaxError: The requested module 'http://localhost:3000/bun:wrap' does not provide 
an export named '__FastRefreshRuntime' (at solid.js:9:1)
```

**Reply by N0tExisting** [reply]
That might mean that it isn't supported yet

**Reply by tonyrewin** [reply]
**react-refresh** should not be a mandatory too!

**Comment by cetorres** [comment]
For me, `bun start` works better. 
But an option to create SolidJS apps like: `bun create solidjs myApp` would be great!

**Comment by Jarred-Sumner** [comment, marked as answer]
Bun has a builtin SolidJS transform (per tweet from a few months ago)

It's honestly like 25% implemented. More work needs to be done on it before it'll be good.

You can try it with `bun dev`, but make sure to run `bun bun ./entry-point.js`

To configure Bun to use the SolidJS transform, `jsxImportSource` should start with `solid-js`:

https://github.com/oven-sh/bun/blob/a1a0c1847805691ec9398eeff6545119a5ad6566/src/options.zig#L910-L914

This can be configured several different ways:
- tsconfig.json/jsconfig.json `jsxImportSource`
- CLI flag: `--jsx-import-source`

- bunfig.toml: `jsx = "solid"`:
https://github.com/oven-sh/bun/blob/a1a0c1847805691ec9398eeff6545119a5ad6566/src/bunfig.zig#L425-L441

**Comment by NyanHelsing** [comment]
<img width="848" alt="Screenshot 2023-04-10 at 1 34 31 AM" src="https://user-images.githubusercontent.com/1523817/230833266-e0ebc13e-9a62-40ad-8fb2-9c51ce7789a3.png">

**Comment by phtdacosta** [comment]
How is the state of Solid JS usage with Bun, so far? Stable enough to give a try on a small project for production @Jarred-Sumner?

**Reply by janos-r** [reply]
I wanted to start a new small project with Solid and Deno or Bun, but both seam to have problems. I wish solid had some of these in their templates. I didn't dive deeper, but seams to me that it is simply not recommended for use right now. I hope I will be wrong soon ^^

**Comment by DaniGuardiola** [comment]
Until Bun has official and complete Solid.js support, you can use my plugin: https://github.com/DaniGuardiola/bun-plugin-solid

**Reply by russelgal** [reply]
> Until Bun has official and complete Solid.js support

it seems like never )

**Reply by DaniGuardiola** [reply]
some day! these folks have a ton of hard work in the backlog, gotta be patient :)

in the meantime my plugin should help. I use it to compile Solid.js for Chrome extensions.
